src/Storages/MergeTree/IMergeTreeReader.cpp

```diff
@@ -138,25 +140,33 @@ void IMergeTreeReader::evaluateMissingDefaults(Block additional_columns, Columns
 {
     try
     {
-        size_t num_columns = requested_columns.size();
+        size_t num_columns = original_requested_columns.size();
 
         if (res_columns.size() != num_columns)
             throw Exception(ErrorCodes::LOGICAL_ERROR, "invalid number of columns passed to MergeTreeReader::fillMissingColumns. "
                             "Expected {}, got {}", num_columns, res_columns.size());
 
-        /// Convert columns list to block.
+        NameSet full_requested_columns_set;
+        NamesAndTypesList full_requested_columns;
+
+        /// Convert columns list to block. And convert subcolumns to full columns.
+        /// Defaults should be executed on full columns to get correct values for subcolumns.
         /// TODO: rewrite with columns interface. It will be possible after changes in ExpressionActions.
-        auto name_and_type = requested_columns.begin();
-        for (size_t pos = 0; pos < num_columns; ++pos, ++name_and_type)
+
+        auto it = original_requested_columns.begin();
+        for (size_t pos = 0; pos < num_columns; ++pos, ++it)
         {
-            if (res_columns[pos] == nullptr)
-                continue;
+            auto name_in_storage = it->getNameInStorage();
 
-            additional_columns.insert({res_columns[pos], name_and_type->type, name_and_type->name});
+            if (full_requested_columns_set.emplace(name_in_storage).second)
+                full_requested_columns.emplace_back(name_in_storage, it->getTypeInStorage());
+
+            if (res_columns[pos])
+                additional_columns.insert({res_columns[pos], it->type, it->name});
```

Comment:
@CurtizJ should this be
```cpp
additional_columns.insert({res_columns[pos], it->type, name_in_storage});
```

Reply:
Almost correct but a bit more complicated: https://github.com/ClickHouse/ClickHouse/pull/70695.